8350209: Preserve adapters in AOT cache

vnkozlov opened the pull request:
[JEP 483](https://bugs.openjdk.org/browse/JDK-8315737) preserves class information in AOT cache which helps Java startup performance.
We should also preserve adapters (i2c, c2i) to avoid their generation during startup.

@ashu-mehra is co-author. He implemented adapters saving in leyden repo.
I did main framework for saving and loading code in AOT cache.

New product flags are introduced:
```
-XX:+StoreAOTCode - store AOT compiled code
-XX:+LoadAOTCode - load AOT compiled code
-XX:AOTCodeMaxSize=10*M - buffer size in bytes for AOT code caching
-XX:+ArchiveAdapters - store adapters and related hash table in AOT cache
```

To use it follow process described in JEP 483:

```
java -XX:AOTMode=record -XX:AOTConfiguration=app.aotconf -cp app.jar App
java -XX:AOTMode=create -XX:AOTConfiguration=app.aotconf -XX:AOTCache=app.aot -XX:+StoreAOTCode -XX:+ArchiveAdapters -cp app.jar
java -XX:AOTCache=app.aot -XX:+LoadAOTCode -cp app.jar App
```

There are several new UL flag combinations to trace the process:

```
-Xlog:aot+codecache+init -Xlog:aot+codecache+exit -Xlog:aot+codecache+stubs
```

Tested tier1-5, stress, xcomp.

vnkozlov commented:
@ashu-mehra I see next failures when we disable AOT code because of VM's environment/settings different and AOT:
```
[0.015s][warning][aot,codecache,init] Disable AOT Code Cache: it was created with CompressedOops::shift() = 3 vs current 0
[0.015s][warning][cds               ] Failed to link AdapterHandlerEntry to its code in the AOT code cache
[0.015s][warning][cds               ] Failed to link AdapterHandlerEntry to its code in the AOT code cache
...
```
Do I miss check somewhere to prevent processing adapters from AOT cache?

vnkozlov commented:
I actually did not find a flag or a value which control loading AOT adapters.

ashu-mehra commented:
> I actually did not find a flag or a value which control loading AOT adapters.

I intentionally did not add a flag to control loading of AOT adapters. The code relies on `load_adapter` returning false if for some reason adapter cannot be loaded from the cache. For instance, in the case you tried where the AOT code cache is disabled, the cache is closed, so the `load_adapter` will immediately return false. I think this follows the design of other APIs to load artifacts from the AOT code cache.
For the messages `Failed to link AdapterHandlerEntry to its code in the AOT code cache` I think we should just change the level from `warning` to `debug`. wdyt?

vnkozlov commented:
> > I actually did not find a flag or a value which control loading AOT adapters.
> 
> I intentionally did not add a flag to control loading of AOT adapters. The code relies on `load_adapter` returning false if for some reason adapter cannot be loaded from the cache. For instance, in the case you tried where the AOT code cache is disabled, the cache is closed, so the `load_adapter` will immediately return false. I think this follows the design of other APIs to load artifacts from the AOT code cache. For the messages `Failed to link AdapterHandlerEntry to its code in the AOT code cache` I think we should just change the level from `warning` to `debug`. wdyt?

First. Since all AOT code flags are diagnostic it would be good to have a flags to switch off patch code loading (stubs, adapters, nmethod). It is not there yet, but I am considering changing `Store|LoadAOTCode` to flag with assigned strings: `+-nmethods,+-stubs, +-adapters` to allow this.

Second. I think we should give such message only if something went wrong when adapters are available and AOT cache is loaded. If AOT code cache is switched of (different configuration or by flag) we should not attempt to link AdapterHandlerEntry to its code in the AOT code. At least for now I can add check `AOTCodeCache::is_on_for_read()`. But I am not sure when. Can you point me the place?

vnkozlov commented:
Note, there is a but somewhere in my code. I see crashes in some tests. I am investigating.

vnkozlov commented:
I am rolling back caching Adapters hash table and fingerprints. Only `load_|store_adapter()` will be used in `Method::make_adapter()` which I am testing now.  Something wrong is going on with cached table.
I want to have clean AOT testing first. Then we can go back to investigate caching table.

ashu-mehra commented:
> Note, there is a but somewhere in my code. I see crashes in some tests. I am investigating.

@vnkozlov can you please share the tests that are failing. I can continue investigating them.

vnkozlov commented:
@ashu-mehra   Here is simple reproducer for linux-x64 and fastdebug local build with current changes.

It reproduces what happens in our AOT testing: we run some Java test to create `jdk.aot` cache file and then use it running tests which includes running `javac` to compile the test.

Second note, current changes do not check `AOTClassLinkning` flag when setting AOT flags. I am not sure if it the cause of issues.  In next update I will add the check (which reduces testing by skipping such cases). 

```
% $MY_JDK/bin/java -XX:-AOTClassLinking -XX:AOTMode=record -XX:AOTConfiguration=app.aotconf -cp hello.jar HelloWorld
Hellow World!
AOTConfiguration recorded: app.aotconf

% $MY_JDK/bin/java -XX:-AOTClassLinking -XX:AOTMode=create -XX:AOTConfiguration=app.aotconf -XX:AOTCache=app.aot -cp hello.jar -XX:ExtraSharedClassListFile=$MY_JDK/lib/classlist
Reading AOTConfiguration app.aotconf and writing AOTCache app.aot
AOTCache creation is complete: app.aot

% $MY_JDK/bin/javac -J-XX:-AOTClassLinking -J-XX:AOTCache=app.aot -cp hello.jar HelloWorld.java

#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000000000000, pid=4035162, tid=4035163
#
# JRE version: Java(TM) SE Runtime Environment (25.0) (fastdebug build 25-internal-LTS-2025-03-21-0238027.vkozlov...)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 25-internal-LTS-2025-03-21-0238027.vkozlov..., mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# j  com.sun.tools.javac.code.Symbol$PackageSymbol.<init>(Lcom/sun/tools/javac/util/Name;Lcom/sun/tools/javac/code/Type;Lcom/sun/tools/javac/code/Symbol;)V+8 jdk.compiler@25-internal
#
```

vnkozlov commented:
It still crash even if I switch off `LoadAOTCode` to prevent loading cached adapters:

```
% $MY_JDK/bin/javac -J-XX:-AOTClassLinking -J-XX:AOTCache=app.aot -J-XX:-LoadAOTCode -cp hello.jar HelloWorld.java

[0.111s][warning][cds] Failed to link AdapterHandlerEntry to its code in the AOT code cache
[0.111s][warning][cds] Failed to link AdapterHandlerEntry to its code in the AOT code cache
...   //// a lot of such warnings
[0.524s][warning][cds] Failed to link AdapterHandlerEntry to its code in the AOT code cache
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000000000000, pid=4036287, tid=4036288
#
# JRE version: Java(TM) SE Runtime Environment (25.0) (fastdebug build 25-internal-LTS-2025-03-21-0238027.vkozlov...)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 25-internal-LTS-2025-03-21-0238027.vkozlov..., mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# j  com.sun.tools.javac.code.Symbol$ModuleSymbol.<init>(Lcom/sun/tools/javac/util/Name;Lcom/sun/tools/javac/code/Symbol;)V+8 jdk.compiler@25-internal
```

vnkozlov commented:
An other issue is missing ResourceMark:

```
#  Internal Error (/workspace/open/src/hotspot/share/memory/resourceArea.cpp:54), pid=406719, tid=406722
#  fatal error: memory leak: allocating without ResourceMark
#
# JRE version: Java(TM) SE Runtime Environment (25.0) (fastdebug build 25-internal-LTS-2025-03-21-1854468.vladimir.kozlov.jdkgit)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 25-internal-LTS-2025-03-21-1854468.vladimir.kozlov.jdkgit, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# V  [libjvm.so+0x1501d2c]  ResourceArea::verify_has_resource_mark() [clone .part.0]+0x2c
....
V  [libjvm.so+0x1501d2c]  ResourceArea::verify_has_resource_mark() [clone .part.0]+0x2c  (resourceArea.cpp:54)
V  [libjvm.so+0x1502128]  (align.hpp:81)
V  [libjvm.so+0x13e6e88]  stringStream::as_string(bool) const+0x58  (ostream.cpp:429)
V  [libjvm.so+0x1566464]  post_adapter_creation(AdapterBlob const*, AdapterHandlerEntry const*)+0x1b4  (sharedRuntime.cpp:2323)
V  [libjvm.so+0x156b234]  AdapterHandlerEntry::restore_unshareable_info(JavaThread*)+0x130  (sharedRuntime.cpp:3034)
V  [libjvm.so+0x12d3bb8]  Method::restore_archived_method_handle_intrinsic(methodHandle, JavaThread*)+0x48  (method.cpp:417)
V  [libjvm.so+0x16731c8]  SystemDictionary::restore_archived_method_handle_intrinsics_impl(JavaThread*)+0xd8  (systemDictionary.cpp:2041)
V  [libjvm.so+0x1673a24]  SystemDictionary::restore_archived_method_handle_intrinsics()+0x74  (systemDictionary.cpp:2028)
```

ashu-mehra commented:
> It reproduces what happens in our AOT testing: we run some Java test to create jdk.aot cache file and then use it running tests which includes running javac to compile the test.

@vnkozlov the reason for the crash is - there is a Method which is not present in CDS archive. In the call to `method::make_adapters()` it locates its `AdapterHandlerEntry` in the `_archived_adapter_handler_table`. However, this code path is missing the code to link this archived adapter to its code in the AOT code cache. So it just returns the unlinked `AdapterHandlerEntry`. Now when this method gets compiled, then the `method::set_code()` sets `_from_interpreted_entry` to `adapter->_i2c_entry` which is null. And so the next invocation of the method from the interpreter results in jumping to a nullptr causing the crash.
I am working on fixing this.

ashu-mehra commented:
@vnkozlov I have pushed some commits on top of the changes made in this PR to fix the issues you mentioned and some other issues I came across. Can you please pull in these changes - https://github.com/openjdk/jdk/compare/master...ashu-mehra:jdk:8350209 and check if it resolves the issues you are facing.

vnkozlov commented:
Thank you, @ashu-mehra. I submitted testing.

vnkozlov commented:
@ashu-mehra my testing passed - no new failures. I integrated your patch.
I will work on AOT Code flags ergonomic settings after Ioi's PR: #24401

vnkozlov commented:
@ashu-mehra, I think next improvements will be saving whole adapters (as you did for nmethods) and preloading them all during startup like we do for "preload" nmethods.

ashu-mehra commented:
>  my testing passed - no new failures. I integrated your patch.

Thanks for testing. I am working on saving whole adapters (and blobs). 

>  preloading them all during startup like we do for "preload" nmethods.

Currently the adapters are loaded on demand as the methods referencing the adapters are loaded from the AOT cache. Preloading them would require some more changes in that logic. I will do that after the changes for saving the whole adapters.

ashu-mehra commented:
@vnkozlov are there any specific tests or any additional options are passed to tests for testing these changes? I can try running them locally before sharing additional code changes.

vnkozlov commented:
> @vnkozlov are there any specific tests or any additional options are passed to tests for testing these changes? I can try running them locally before sharing additional code changes.

I did not tried it myself. @iklam should know how to run it locally. But based on comment in [TestSetupAOT.java](https://github.com/openjdk/jdk/blob/master/test/setup_aot/TestSetupAOT.java#L36) you should specify `JTREG=AOT_JDK=true`

iklam commented:
> > @vnkozlov are there any specific tests or any additional options are passed to tests for testing these changes? I can try running them locally before sharing additional code changes.
> 
> I did not tried it myself. @iklam should know how to run it locally. But based on comment in [TestSetupAOT.java](https://github.com/openjdk/jdk/blob/master/test/setup_aot/TestSetupAOT.java#L36) you should specify `JTREG=AOT_JDK=true`

I think running a good subset of the runtime tests with the following should provide good coverage.

```
make test JTREG=AOT_JDK=true TEST=open/test/hotspot/jtreg/runtime/<something>
```

In our internal testing, we run many of the tests under `open/test` in this mode.

Also, I am planning to upstream the these tests from leyden to mainline. That should increase coverage. I think you can try copying them into your mainline repo and remove the `id=leyden` part. They might already work with mainline.

https://github.com/openjdk/leyden/tree/premain/test/hotspot/jtreg/runtime/cds/appcds/applications

ashu-mehra commented:
@iklam 

> I think running a good subset of the runtime tests with the following should provide good coverage.
> 
> make test JTREG=AOT_JDK=true TEST=open/test/hotspot/jtreg/runtime/<something>
> In our internal testing, we run many of the tests under open/test in this mode.

Can we make this, i.e. running with JTREG=AOT_JDK, work for premain branch as well? Or is this already available in premain?

iklam commented:
> @iklam
> 
> > I think running a good subset of the runtime tests with the following should provide good coverage.
> > make test JTREG=AOT_JDK=true TEST=open/test/hotspot/jtreg/runtime/
> > In our internal testing, we run many of the tests under open/test in this mode.
> 
> Can we make this, i.e. running with JTREG=AOT_JDK, work for premain branch as well? Or is this already available in premain?

Yes, this feature has been merged into the premain branch.

ashu-mehra commented:
@vnkozlov I have added a couple of more commits (see my branch https://github.com/openjdk/jdk/compare/master...ashu-mehra:jdk:8350209) to store the CodeBlob as is into the AOT code cache. So we now don't need separate APIs to store/load different types of code blobs. The new API can be used to cater to the existing use-cases to store adapters and exception blob. What do you think about these changes? https://github.com/openjdk/jdk/commit/3fce9a5b1d57028fa15f6dd630f850ecb170473f and https://github.com/openjdk/jdk/commit/994cee0dd4070888bed2096c32b7bde8c39014b0
If they look fine can you please integrate them and run tests.

vnkozlov commented:
Yes, this looks fine. I may modify few places and then will test it. Thank you, for doing it.

vnkozlov commented:
@ashu-mehra  Can you rebase your patch to latest state of this branch. I merged mainline and I can't apply your patch.

vnkozlov commented:
/contributor add @ashu-mehra

ashu-mehra commented:
@vnkozlov I have rebased my patches to the current state. They are available here: https://github.com/openjdk/jdk/compare/master...ashu-mehra:jdk:8350209

ashu-mehra commented:
@iklam I tested my changes locally on x86-64 using following command:

```make test JTREG=AOT_JDK=true TEST=test/hotspot/jtreg/runtime CONF=...```

and a lot of tests failed with this error message:

```Option AOTCache cannot be used at the same time with -Xshare:on, -Xshare:auto, -Xshare:off, -Xshare:dump, DumpLoadedClassList, SharedClassListFile, or SharedArchiveFile```

It seems many of the tests run with -Xshare:dump option. I guess this is expected, right?

iklam commented:
> @iklam I tested my changes locally on x86-64 using following command:
> 
> `make test JTREG=AOT_JDK=true TEST=test/hotspot/jtreg/runtime CONF=...`
> 
> and a lot of tests failed with this error message:
> 
> `Option AOTCache cannot be used at the same time with -Xshare:on, -Xshare:auto, -Xshare:off, -Xshare:dump, DumpLoadedClassList, SharedClassListFile, or SharedArchiveFile`
> 
> It seems many of the tests run with -Xshare:dump option. I guess this is expected, right?

I double checked our test setup and we run with `JTREG=AOT_JDK=true TEST=test/hotspot/jtreg/:hotspot_runtime_no_cds`. This skips all CDS tests that use the above options.

ashu-mehra commented:
> @vnkozlov I have rebased my patches to the current state. They are available here: https://github.com/openjdk/jdk/compare/master...ashu-mehra:jdk:8350209

I have pushed more changes to fix compilation failures.

vnkozlov commented:
New test `runtime/cds/appcds/aotClassLinking/AOTCacheWithZGC.java` failed with:
```
#  fatal error: Address 0x00007f6ff6770950 for runtime target 'ZBarrierSetRuntime::load_barrier_on_phantom_oop_field_preloaded(oopDesc*, oop*)+0' is missing in AOT Code Cache addresses table
```

I will add it and rerun testing.

The test used next flags:

```
-XX:+UseZGC -XX:AOTMode=create -XX:AOTConfiguration=AOTCacheWithZGC.aotconfig -XX:AOTCache=AOTCacheWithZGC.aot
```

Call stack:
```
V  [libjvm.so+0x65419e]  AOTCodeAddressTable::id_for_address(unsigned char*, RelocIterator, CodeBlob*) [clone .part.0]+0x5fe  (aotCodeCache.cpp:1298)
V  [libjvm.so+0x65497d]  AOTCodeCache::write_loadtime_relocations(CodeBlob&)+0x5cd  (aotCodeCache.cpp:1259)
V  [libjvm.so+0x6554cf]  AOTCodeCache::store_code_blob(CodeBlob&, AOTCodeEntry::Kind, unsigned int, char const*, int, int*)+0x42f  (aotCodeCache.cpp:738)
V  [libjvm.so+0x185acd7]  AdapterHandlerLibrary::generate_adapter_code(AdapterBlob*&, AdapterHandlerEntry*, int, BasicType*, bool)+0x4c7  (sharedRuntime.cpp:2865)
V  [libjvm.so+0x185ad4b]  AdapterHandlerLibrary::create_adapter(AdapterBlob*&, AdapterFingerPrint*, int, BasicType*, bool)+0x6b  (sharedRuntime.cpp:2883)
V  [libjvm.so+0x185cfd3]  AdapterHandlerLibrary::initialize()+0x1d3  (sharedRuntime.cpp:2565)
V  [libjvm.so+0xf79062]  init_globals()+0x92  (init.cpp:159)
```

vnkozlov commented:
An other failure in next tests on Mac M1:
vmTestbase/nsk/jvmti/GenerateEvents/genevents001/TestDescription.java
vmTestbase/nsk/jvmti/scenarios/events/EM04/em04t001/TestDescription.java
```
#  Internal Error (/workspace/open/src/hotspot/share/prims/jvmtiExport.hpp:467), pid=30154, tid=26887
#  assert(name != nullptr) failed: all code blobs must be named
...
V  [libjvm.dylib+0x59306c]  print_error_for_unit_test(char const*, char const*, char*)+0x0
V  [libjvm.dylib+0xb56828]  CodeBlobCollector::do_vtable_stub(VtableStub*)+0x0
V  [libjvm.dylib+0x4be8b0]  CodeCache::blobs_do(void (*)(CodeBlob*))+0xdc
V  [libjvm.dylib+0xb56ab0]  CodeBlobCollector::collect()+0x1dc
V  [libjvm.dylib+0xb56b44]  JvmtiCodeBlobEvents::generate_dynamic_code_events(JvmtiEnv*)+0x50
```

Which should not happen since [do_vtable_stub](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/prims/jvmtiCodeBlobEvents.cpp#L149) passes C string:
```
    JvmtiCodeBlobDesc* scb = new JvmtiCodeBlobDesc(vs->is_vtable_stub() ? "vtable stub" : "itable stub",
                                                   vs->code_begin(), vs->code_end());
```

But you did modified how blobs are created. And another call to `JvmtiCodeBlobDesc` use `desc->name()`: [jvmtiCodeBlobEvents.cpp#L179](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/prims/jvmtiCodeBlobEvents.cpp#L179)

The failure happens when AOTCache is loaded. So we reading adapters. I reproduced failure locally with:
```
make test JTREG=AOT_JDK=true TEST=vmTestbase/nsk/jvmti/GenerateEvents/genevents001/TestDescription.java
```

vnkozlov commented:
An other interesting thing about this failure is that we have only few adapters:
```
CodeCache: size=245760Kb, used=1790Kb, max_used=1798Kb, free=243967Kb
 total_blobs=863, nmethods=288, adapters=7, full_count=0
```

vnkozlov commented:
@ashu-mehra  I see this in your changes:
```
+void CodeBlob::prepare_for_archiving_impl() {
+  set_name(nullptr);
```

But I don't see where you restore name.

vnkozlov commented:
> An other interesting thing about this failure is that we have only few adapters:
> 
> ```
> CodeCache: size=245760Kb, used=1790Kb, max_used=1798Kb, free=243967Kb
>  total_blobs=863, nmethods=288, adapters=7, full_count=0
> ```

May be restored adapters are not counted as adapters.   Running with `-Xlog:aot+codecache+stubs` shows that we read adapters.

vnkozlov commented:
I printed blob which has no name:
```
[CodeBlob (0x000000013759e808)]
Framesize: 0
BufferBlob (0x000000013759e808) used for (null)
```

vnkozlov commented:
I fixed issue with name (passed it into `CodeBlob::restore()` to set) and updated adapters count in `CodeBlob::create()`. There is new crash in `AOTCacheWithZGC.java` which I am investigating. I think we are missing some runtime address in AOT address table:
```
#  Internal Error (/workspace/open/src/hotspot/share/gc/shared/oopStorageSet.cpp:107), pid=1129456, tid=1129459
#  assert(_storages[index] != nullptr) failed: oopstorage_init not yet called
...
V  [libjvm.so+0x169ea2c]  OopStorageSet::print_containing(void const*, outputStream*)+0x44c  (oopStorageSet.cpp:107)
V  [libjvm.so+0x16a2cdd]  os::print_location(outputStream*, long, bool)+0x1ed  (os.cpp:1363)
V  [libjvm.so+0x6542a4]  AOTCodeAddressTable::id_for_address(unsigned char*, RelocIterator, CodeBlob*) [clone .part.0]+0x624  (aotCodeCache.cpp:1308)
V  [libjvm.so+0x6548d9]  AOTCodeCache::write_loadtime_relocations(CodeBlob&)+0x449  (aotCodeCache.cpp:1267)
V  [libjvm.so+0x6555af]  AOTCodeCache::store_code_blob(CodeBlob&, AOTCodeEntry::Kind, unsigned int, char const*, int, int*)+0x42f  (aotCodeCache.cpp:743)
V  [libjvm.so+0x185adb7]  AdapterHandlerLibrary::generate_adapter_code(AdapterBlob*&, AdapterHandlerEntry*, int, BasicType*, bool)+0x4c7  (sharedRuntime.cpp:2865)
V  [libjvm.so+0x185ae2b]  AdapterHandlerLibrary::create_adapter(AdapterBlob*&, AdapterFingerPrint*, int, BasicType*, bool)+0x6b  (sharedRuntime.cpp:2883)
V  [libjvm.so+0x185d0b3]  AdapterHandlerLibrary::initialize()+0x1d3  (sharedRuntime.cpp:2565)
V  [libjvm.so+0xf79142]  init_globals()+0x92  (init.cpp:159)
```

vnkozlov commented:
Finally "clean" testing. @ashu-mehra I pushed your patch with few fixes and renames.

ashu-mehra commented:
@vnkozlov Re renaming `CDSAccess` to `AOTCodeAccess` I feel it should have been `AOTCacheAccess`

vnkozlov commented:
I am testing flags ergonomic now. I will rename AOTCodeAccess after I am done with that.

ashu-mehra commented:
I was wondering why I didn't hit that invalid assert in `AOTCodeCache::compile_code_blob`:
```
assert(stored_count == AdapterHandlerEntry::ENTRIES_COUNT, "wrong caller expectations");
```
Turns out during assembly phase we run in interpreter mode.

https://github.com/openjdk/jdk/blob/9a3f9997b68a1f64e53b9711b878fb073c3c9b90/src/hotspot/share/cds/cdsConfig.cpp#L519-L527 

The `mode_flag_cmd_line` is false here.
And this prevents generation of exception blob used by c2 compiler.
I guess this is fine for now as we are not storing any nmethod in the AOT Code cache.
But eventually we will have to enable compilation in assembly phase as well.

For the record, in `premain` we have additional condition that avoids tweaking execution mode when dumping final image:

```
  if (is_dumping_static_archive()) {
    if (is_dumping_preimage_static_archive() || is_dumping_final_static_archive()) {
      // Don't tweak execution mode
    } else if (!mode_flag_cmd_line) {
      // By default, -Xshare:dump runs in interpreter-only mode, which is required for deterministic archive.
      //
      // If your classlist is large and you don't care about deterministic dumping, you can use
      // -Xshare:dump -Xmixed to improve dumping speed.
      Arguments::set_mode_flags(Arguments::_int);
```

ashu-mehra commented:
@vnkozlov I have pushed another change to link archived adapters during startup instead of doing it on-demand. Please see this commit: https://github.com/openjdk/jdk/commit/7b983e1e2a39e6cc46ce8764355372c3fde371bf

vnkozlov commented:
@ashu-mehra, yes I noticed it when I start this changes. I am passing `TEST_OPTS_JAVA_OPTIONS="-Xmixed` to `make test` when running AOT testing to test caching of C2 blob.

I did not notice that difference with `premain`.

@iklam   can we do the same in mainline? I mean don't enforce `-Xint` when `is_dumping_final_static_archive()`.

vnkozlov commented:
Can you do performance runs and compare (on any hardware)?

vnkozlov commented:
Is this patch in `premain` repo or it is new code?

Can you describe here what it does and when?

iklam commented:
> @ashu-mehra, yes I noticed it when I start this changes. I am passing `TEST_OPTS_JAVA_OPTIONS="-Xmixed` to `make test` when running AOT testing to test caching of C2 blob.
> 
> I did not notice that difference with `premain`.
> 
> @iklam can we do the same in mainline? I mean don't enforce `-Xint` when `is_dumping_final_static_archive()`.

Yes, we should do that. -Xint is really intended for producing deterministic default CDS archives during the JDK build, but we don't use `is_dumping_final_static_archive()` when dumping the default archives.

vnkozlov commented:
I decided to push your last patch. We can roll it back if we find an issue.
I also added `is_dumping_final_static_archive()` check to not tweak execution mode as discussed.

ashu-mehra commented:
> Is this patch in premain repo or it is new code?
> 
> Can you describe here what it does and when?

This patch is not in premain. Its new code. Before this patch an archived AdapterHandlerEntry was linked to its code in AOT code cache when the AdapterHandlerEntry is "adopted" by the JVM from the AOT Cache, which would happen when the first method refering to the AdapterHandlerEntry is restored from the AOT cache.
This new code now links all the archived AdapterHandlerEntry-s upfront during AdapterHandlerLibrary initialization. So by the time methods are restored, their AdapterHandlerEntry have already been linked.

I don't expect any measurable performance improvement. But I do think that this is a better approach because now we can be sure that the AdapterHandlerEntry-s are always linked before they are used. There shouldn't be any scenario where we miss linking AdapterHandlerEntry, which happened when it was done on-demand. It also somewhat simplifies the code.

I actually did this patch based on your comment [here](https://github.com/openjdk/jdk/pull/24158#issuecomment-2776984497) :

>  I think next improvements will be saving whole adapters (as you did for nmethods) and preloading them all during startup like we do for "preload" nmethods.

I though this is what you meant by "preloading them all during startup". But based on the meeting yesterday, I understand you were referring to "bulk loading of the adapters" instead of doing them one blob at a time.

vnkozlov commented:
Thank you @ashu-mehra for explanation. Simpler code wins even if there is no performance change.

Yes, I was talking about bulk loading of all archived adapters. And this change goes in that direction. Good.

ashu-mehra commented on src/hotspot/share/cds/aotCacheAccess.cpp:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2023, 2025, Oracle and/or its affiliates. All rights reserved.
```
Shouldn't the copyright year be just 2025? Same for other new files.

vnkozlov replied:
Right. I will fix that.

ashu-mehra commented on src/hotspot/share/cds/cdsConfig.cpp:
```diff
@@ -847,3 +852,22 @@ bool CDSConfig::is_dumping_method_handles() {
+// This affects only JITed code because it may have embedded oops and metadata pointers
+// which AOT code encodes as offsets in final CDS archive regions.
+
+static bool _is_dumping_aot_code_enabled = false;
```
Can we rename it to `_is_dumping_aot_code`? IMO the `_enabled` suffix looks redundant here.

vnkozlov replied:
`_is_dumping_aot_code` means "we are dumping code now" for me.
`_is_dumping_aot_code_enabled` means we are allowed to dump code now. Which is correct because it will be used during attempt to dump nmethods:
```
AOTCodeEntry* AOTCodeCache::store_nmethod(nmethod* nm, AbstractCompiler* compiler, bool for_preload) {
  if (!CDSConfig::is_dumping_aot_code_enabled()) {
    return nullptr; // The metadata and heap in the CDS image haven't been finalized yet.
  }
```

ashu-mehra commented on src/hotspot/share/cds/cds_globals.hpp:
```diff
@@ -127,6 +127,27 @@
                                                                             \
+  /* AOT Code flags */                                                      \
+                                                                            \
+  product(bool, AOTCodeCaching, false, DIAGNOSTIC,                          \
```
This PR is not using `AOTCodeCaching` flag. Does it make sense to add it later when support for nmethods is added?

vnkozlov replied:
Yes.

Originally I used it as main flag to disable all code caching including stubs and adapters. But now it will only control JITed AOT code.

ashu-mehra commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -0,0 +1,1368 @@
+  for (int i = 0; i < stored_count; i++) {
+    uint32_t off = *(uint32_t*)addr(offset);
+    offset += sizeof(uint32_t);
+    const char* entry_name = AdapterHandlerEntry::entry_name(i);
```
This is Adapter specific code. But we use this API for blobs as well. I thing we should either remove it or refactor to make it more generic.

vnkozlov replied:
I specialized `entry_name` for adapters otherwise "".

ashu-mehra commented on src/hotspot/share/runtime/sharedRuntime.cpp:
```diff
@@ -2612,110 +2741,290 @@ AdapterHandlerEntry* AdapterHandlerLibrary::get_adapter(const methodHandle& meth
+  assert(_linked, "AdapterHandlerEntry must now be linked");
+}
+
+void AdapterHandlerLibrary::link_aot_adapters() {
```
I am wondering if we need to acquire `AdapterHandlerLibrary_lock`, as done by other methods called during this stage (eg `lookup_simple_adapters` and `AdapterHandlerLibrary::initialize`). 
Linking of adapters happens very early in bootstrap, where we have only one thread. But this lock was acquired in earlier code version of the code during bootstrap as well. So to keep things consistent, I feel we should hold the lock here.

vnkozlov replied:
Please, give me patch where we should place locks.

ashu-mehra replied:
@vnkozlov here is the patch for adding the lock: https://github.com/ashu-mehra/jdk/commit/a1c463ddc85e7c179b32ce1c3f209f623cb6db71
It also adds back PerfTraceTimer that I missed in earlier refactoring.

ashu-mehra commented:
@vnkozlov I have another [patch](https://github.com/ashu-mehra/jdk/commit/4982ef91df66407ef9ec2defa732045c7f2fc82c) that simplifies string handling a bit. Main idea is to ensure strings in `_C_strings` array are unique when added through `AOTCodeCache::add_C_string`, and that allows us to avoid some processing later on. And the hash-es are not required any more in the AOT code cache. So some marginal space saving. wdyt about this?

vnkozlov commented:
> @vnkozlov In addition to above comments I have made some minor fixes and changes on top of this PR. They can be seen here: [ashu-mehra@004711c](https://github.com/ashu-mehra/jdk/commit/004711c6a02495a154cd6970f8ab384b1df83a7f) Feel free to merge them in this PR if they are appropriate.

I will use it. The only thing I don't want in our final PR is enabling stubs caching (`exception_blob` only now). I am actually thinking about removing all stubs related code. What do you think?

vnkozlov commented:
> @vnkozlov I have another [patch](https://github.com/ashu-mehra/jdk/commit/4982ef91df66407ef9ec2defa732045c7f2fc82c) that simplifies string handling a bit. Main idea is to ensure strings in `_C_strings` array are unique when added through `AOTCodeCache::add_C_string`, and that allows us to avoid some processing later on. And the hash-es are not required any more in the AOT code cache. So some marginal space saving. wdyt about this?

Nice! I will test it and push.

ashu-mehra commented:
> I will use it. The only thing I don't want in our final PR is enabling stubs caching (exception_blob only now). I am actually thinking about removing all stubs related code. What do you think?

Oh I though that was a bug. If you plan not to enable stubs caching, I am fine with that.
btw @adinn and I had a discussion other day to try resurrect our stubs and blobs prototype (which was done on premain) on top of this PR. If we are able to make it work, then we can enable stubs caching as part of that work.

vnkozlov commented:
> > I will use it. The only thing I don't want in our final PR is enabling stubs caching (exception_blob only now). I am actually thinking about removing all stubs related code. What do you think?
> 
> Oh I though that was a bug. If you plan not to enable stubs caching, I am fine with that. btw @adinn and I had a discussion other day to try resurrect our stubs and blobs prototype (which was done on premain) on top of this PR. If we are able to make it work, then we can enable stubs caching as part of that work.

I filed https://bugs.openjdk.org/browse/JDK-8354887 and assigned @adinn to it.

vnkozlov commented:
> I filed https://bugs.openjdk.org/browse/JDK-8354887 and assigned @adinn to it.

Note, I am not targeting it for JDK 25 (unless you make it).  May be next JDK 26 release if it will be ready. I will work on nmethods caching in parallel.

vnkozlov commented:
I ran HelloWorld on big  linux-x64 machine like next and used `perf` to get numbers:
```
Generating adapters:

(perf stat -r 100 ./build/product/images/jdk/bin/java -Xms128m -XX:ParallelGCThreads=2 -XX:CICompilerCount=2 -XX:+UseParallelGC -XX:AOTCache=app.aotcache -XX:+UnlockDiagnosticVMOptions -XX:-AOTAdapterCaching -cp hello.jar HelloWorld  > /dev/null) 2>&1 | grep elapsed
         0.0246760 +- 0.0000376 seconds time elapsed  ( +-  0.15% )

With using caching adapter:

(perf stat -r 100 ./build/product/images/jdk/bin/java -Xms128m -XX:ParallelGCThreads=2 -XX:CICompilerCount=2 -XX:+UseParallelGC -XX:AOTCache=app.aotcache -XX:+UnlockDiagnosticVMOptions -XX:+AOTAdapterCaching -cp hello.jar HelloWorld  > /dev/null) 2>&1 | grep elapsed
         0.0240801 +- 0.0000308 seconds time elapsed  ( +-  0.13% )
```

It is 2.4% improvement!!!

vnkozlov commented:
With default flags (`java -cp hello.jar HelloWorld`) I got 5.7%!!! : 
```
0.0316486 +- 0.0000673
and
0.0298249 +- 0.0000332
```

vnkozlov commented:
After more thinking I decide to remove stubs caching from these changes. It was nice to have it for testing but for final patch I don't think we should include it. You can restore it in your local repo by reversing removal.

vnkozlov commented:
@ashu-mehra  I got crash running `compiler/startup/StartupOutput.java` test with small CodeCache (tier1 normal execution - no AOT):
```
----------System.err:(33/2013)----------
 stdout: [[0.441s][warning][codecache] CodeCache is full. Compiler has been disabled.
[0.441s][warning][codecache] Try increasing the code cache size using -XX:ReservedCodeCacheSize=
CodeCache: size=1280Kb used=1277Kb max_used=1277Kb free=2Kb
 bounds [0x00000228ea120000, 0x00000228ea260000, 0x00000228ea260000]
 total_blobs=249, nmethods=0, adapters=183, full_count=1
Compilation: disabled (not enough contiguous free space left), stopped_count=1, restarted_count=0
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x00007ffc6a4855cd, pid=4092, tid=32740
#
# JRE version: Java(TM) SE Runtime Environment (25.0) (fastdebug build 25-internal-LTS-2025-04-16-2049387.vladimir.kozlov.jdkgit)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 25-internal-LTS-2025-04-16-2049387.vladimir.kozlov.jdkgit, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, windows-amd64)
# Problematic frame:
# V  [jvm.dll+0x3555cd]Error occurred during initialization of boot layer
java.lang.OutOfMemoryError: Out of space in CodeCache for adapters
];
 stderr: [Java HotSpot(TM) 64-Bit Server VM warning: CodeCache is full. Compiler has been disabled.
Java HotSpot(TM) 64-Bit Server VM warning: Try increasing the code cache size using -XX:ReservedCodeCacheSize=
]
 exitValue = 1

java.lang.RuntimeException: '# A fatal error' found in stdout
	at jdk.test.lib.process.OutputAnalyzer.stdoutShouldNotContain(OutputAnalyzer.java:337)
```

Unfortunately hs_err file was not saved.  We should not crash but throw exception, something like:

```
Java HotSpot(TM) 64-Bit Server VM warning: CodeCache is full. Compiler has been disabled.
Java HotSpot(TM) 64-Bit Server VM warning: Try increasing the code cache size using -XX:ReservedCodeCacheSize=
java.lang.reflect.InvocationTargetException
...
Caused by: java.lang.VirtualMachineError: Out of space in CodeCache for adapters
```

  Can you check your changes?

vnkozlov commented:
```
For random generator using seed: 7776048711032458590
To re-run test with same seed value please add "-Djdk.test.lib.random.seed=7776048711032458590" to command line.
```

ashu-mehra commented:
I did hit `java.lang.OutOfMemoryError: Out of space in CodeCache for adapters` but didn't get a crash. I have been trying this on linux. Could this be specific to Windows? I don't have access to a windows box to try this out.

vnkozlov commented:
Final PR: https://github.com/openjdk/jdk/pull/24740
I am closing this one.
